cli/internal/core/engine.go

```
@@ -385,11 +412,131 @@ func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) erro
+				Exclusions: taskDef.Outputs.Exclusions,
+			}
+		}
+
+		mergedTaskDefinition.ShouldCache = taskDef.ShouldCache
+		if bookkeepingTaskDef.HasField("EnvVarDependencies") {
+			mergedTaskDefinition.EnvVarDependencies = taskDef.EnvVarDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("TopologicalDependencies") {
+			mergedTaskDefinition.TopologicalDependencies = taskDef.TopologicalDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("TaskDependencies") {
+			mergedTaskDefinition.TaskDependencies = taskDef.TaskDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("Inputs") {
+			mergedTaskDefinition.Inputs = taskDef.Inputs
+		}
+
+		if bookkeepingTaskDef.HasField("OutputMode") {
+			mergedTaskDefinition.OutputMode = taskDef.OutputMode
+		}
+		mergedTaskDefinition.Persistent = taskDef.Persistent
+	}
+
+	// Set defaults if we had no config for this.
+	if mergedTaskDefinition.Outputs == nil {
+		mergedTaskDefinition.Outputs = &fs.TaskOutputs{
+			Inclusions: []string{},
+			Exclusions: []string{},
+		}
+	}
+
+	return mergedTaskDefinition, nil
+}
+
+func (e *Engine) getTaskDefinitionChain(rootPipeline *fs.Pipeline, pkg *fs.PackageJSON, taskID string, taskName string) ([]fs.BookkeepingTaskDefinition, error) {
+	// Look for the taskDefinition in the root pipeline. We'll wait to throw errors until the end
+	rootTaskDefinition, _ := rootPipeline.GetTask(taskID, taskName)
+	var workspaceDefinition *fs.BookkeepingTaskDefinition
+
+	// If the taskID is a root task (e.g. //#build), we don't need to look
+	// for a workspace task, since these can only be defined in the root turbo.json.
+	taskIDPackage, _ := util.GetPackageTaskFromId(taskID)
+	if taskIDPackage != util.RootPkgName && taskIDPackage != ROOT_NODE_NAME {
+		// Look up task definition in turbo.json in the workspace directory
+		workspaceConfigPath := turbopath.AbsoluteSystemPath(pkg.Dir).UntypedJoin("turbo.json")
+		workspaceTurboJSON, _ := fs.ReadTurboConfig(workspaceConfigPath)
```

I think we only want to swallow a "file not found" error and throw the others. It might be confusing to the user if they have a bad `apps/a/turbo.json` that defines a task and they only get a "Could not find "task" in root turbo.json or "a" workspace.

---

Good call. Added this to the deferrals list. I need to work on using the cached set of turbo.jsons, i'll update the messaging and error handling while i'm there.